core/dbt/parser/manifest.py

```diff
@@ -1384,24 +1382,85 @@ def _process_refs_for_node(manifest: Manifest, current_project: str, node: Manif
     target_model_package: Optional[str] = None
-    if node.resource_type == "model" and len(node.relationships) > 0:
-        for relationship in node.relationships:
+    
+    """Limit this parsing to public models/seeds that have relationships"""
+    if node.resource_type in ("model","seed") and node.is_public == True:
+
+        """Creating a list that will be populated with models missing the in_public
+        flag that are being referenced with relationships"""
+        non_public_models = []
+
+        """Setting the primary_keys field in the node to be a combination of all 
+        columns that have is_primary_key set to true"""
+        primary_key_columns = [
+            column for column in node.columns.values() if column.is_primary_key
+        ]
+        """Appending the values if they don't exist just in case someone manually 
+        tries to set the primary key at the model level"""
+        for column in primary_key_columns:
+            if column.name not in node.primary_keys:
+                node.primary_keys.append(column.name)
+
+        """Use a generator expression to create the set of public models/seeds
+        that have relationships, which we'll use to iterate on"""
+        nodes_with_relationships = (relationship for relationship in node.relationships if len(node.relationships) > 0)
+        for relationship in nodes_with_relationships:
+
+            """Here we overwrite the base value of the join_keys if it is a string and replace
+            it with a list value for the manifest"""
+            if type(relationship.join_keys) == str:
+                relationship.join_keys = [relationship.join_keys]
+
             to_model = manifest.resolve_ref(
                 relationship.to, target_model_package, current_project, node.package_name
             )
+
+            """Create a list of models that have relationships pointed at them that
+            have not been set to is_public. Use this list to raise CompilationException"""
+            if to_model.is_public == False:
+                non_public_models.append(to_model.name)
+
+            """Using the to_model, create an inverse relationship in the model"""
             inverse_relationship = EntityRelationship(
                 to=node.name,
-                join_key=relationship.join_key,
+                join_keys=relationship.join_keys,
                 relationship_type=EntityRelationshipType(relationship.relationship_type.inverse()),
             )
-            to_model.relationships.append(inverse_relationship)
-            manifest.update_node(to_model)
+            
+            # TODO: Clean up this very messy code. See if way we can check 
+            # relationship.to against to_model.relationships without it breaking 
+            # because the latter is a EntityRelationship
+            """Checks if the relationship already exists and only creates if not"""
```

Comment:
this is good! can you drop (maybe in notion) a note about all the conditions we're checking here, just so we have it documented?

Reply:
Resolved